cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/ProgramLifecycleService.java

```diff
@@ -436,6 +463,34 @@ private boolean isRunning(ProgramId programId) throws Exception {
     return ProgramStatus.STOPPED != getProgramStatus(programId);
   }
 
+  /**
+   * Returns the program status.
+   * @param programId the id of the program for which the running status in all versions of the app is found
+   * @return whether the same program is running in any versions of the app
+   * @throws NotFoundException if the application to which this program belongs was not found
+   */
+  private boolean isRunningInSameProgram(ProgramId programId) throws Exception {
+    // check that app exists
+    Collection<ApplicationId> appIds = store.getAllAppVersionsAppIds(programId.getNamespace(),
+                                                                     programId.getApplication());
+    if (appIds == null) {
+      throw new NotFoundException(Ids.namespace(programId.getNamespace()).app(programId.getApplication()).toId());
+    }
+    ApplicationSpecification appSpec = store.getApplication(programId.getParent());
+    for (ApplicationId appId : appIds) {
+      ProgramId pId = appId.program(programId.getType(), programId.getProgram());
+      if (getExistingAppProgramStatus(appSpec, pId).equals(ProgramStatus.RUNNING)) {
+        return true;
+      }
+    }
+    return false;
+  }
+
+  private boolean isCocurrentRunsInSameAppForbidden(ProgramType type) {
+    // Concurrent runs are only allowed for the Workflow and MapReduce
+    return EnumSet.of(ProgramType.WORKER, ProgramType.FLOW).contains(type);
```

Comment:
Hmm, we decided to retain the same behavior as in apps. That we didn't have concurrent runs for Workers in apps.